### ISSUE TITLE
Extracting an `Option` chooses the wrong type for `self`

### ISSUE THREAD

**Issue by deech** [nim-lang/Nim#20309]
### What happened?

I hit this bug when I had a parameterized object of `Option` procs. A minimal failing case is:

```nim
import std/[options]
type
  F1*[T] = proc(self:T,callback:proc(self:T))
  F2*[T] = proc(self:T,callback:proc(self:T))
  F3*[T] = proc(self:T,callback:proc(self:T))

type
  Functions*[T] = object
    a*: Option[F1[T]]
    b*: Option[F1[T]]
    c*: Option[F1[T]]
    d*: Option[F1[T]]
    e*: Option[F1[T]]
    f*: Option[F1[T]]
    g*: Option[F1[T]]
    h*: Option[F1[T]]
    i*: Option[F2[T]]
    j*: Option[F3[T]]

var o : Functions[int]

static:
  echo typeof(o.j.get)
```

When compiled with a temporary compiler:

```nim
koch temp c  test.nim
```

I get:

```
/home/deech/Nim/test.nim(23, 18) template/generic instantiation of `get` from here
/home/deech/Nim/lib/pure/options.nim(231, 14) Error: type mismatch: got 'F2[system.int]' for 'self.val' but expected 'F3[system.int]'
Traceback (most recent call last)
/home/deech/Nim/compiler/nim.nim(143) nim
/home/deech/Nim/compiler/nim.nim(98) handleCmdLine
/home/deech/Nim/compiler/main.nim(285) mainCommand
/home/deech/Nim/compiler/main.nim(254) compileToBackend
/home/deech/Nim/compiler/main.nim(110) commandCompileToC
/home/deech/Nim/compiler/modules.nim(136) compileProject
/home/deech/Nim/compiler/modules.nim(56) compileModule
/home/deech/Nim/compiler/passes.nim(183) processModule
/home/deech/Nim/compiler/passes.nim(76) processTopLevelStmt
/home/deech/Nim/compiler/sem.nim(656) myProcess
/home/deech/Nim/compiler/sem.nim(624) semStmtAndGenerateGenerics
/home/deech/Nim/compiler/semstmts.nim(2466) semStmt
/home/deech/Nim/compiler/semexprs.nim(1088) semExprNoType
/home/deech/Nim/compiler/semexprs.nim(3103) semExpr
/home/deech/Nim/compiler/semstmts.nim(2408) semStmtList
/home/deech/Nim/compiler/semexprs.nim(3155) semExpr
/home/deech/Nim/compiler/semstmts.nim(2361) semStaticStmt
/home/deech/Nim/compiler/semstmts.nim(2466) semStmt
/home/deech/Nim/compiler/semexprs.nim(1088) semExprNoType
/home/deech/Nim/compiler/semexprs.nim(3103) semExpr
/home/deech/Nim/compiler/semstmts.nim(2408) semStmtList
/home/deech/Nim/compiler/semexprs.nim(2999) semExpr
/home/deech/Nim/compiler/semexprs.nim(2404) semMagic
/home/deech/Nim/compiler/semexprs.nim(1069) semDirectOp
/home/deech/Nim/compiler/semexprs.nim(898) semOverloadedCallAnalyseEffects
/home/deech/Nim/compiler/semcall.nim(574) semOverloadedCall
/home/deech/Nim/compiler/semcall.nim(373) resolveOverloads
/home/deech/Nim/compiler/semcall.nim(95) pickBestCandidate
/home/deech/Nim/compiler/sigmatch.nim(2571) matches
/home/deech/Nim/compiler/sigmatch.nim(2504) matchesAux
/home/deech/Nim/compiler/sigmatch.nim(2300) prepareOperand
/home/deech/Nim/compiler/semexprs.nim(55) semOperand
/home/deech/Nim/compiler/semexprs.nim(2999) semExpr
/home/deech/Nim/compiler/semexprs.nim(2290) semMagic
/home/deech/Nim/compiler/semmagic.nim(30) semTypeOf
/home/deech/Nim/compiler/semexprs.nim(87) semExprWithType
/home/deech/Nim/compiler/semexprs.nim(71) semExprCheck
/home/deech/Nim/compiler/semexprs.nim(2956) semExpr
/home/deech/Nim/compiler/semexprs.nim(2998) semExpr
/home/deech/Nim/compiler/semexprs.nim(1069) semDirectOp
/home/deech/Nim/compiler/semexprs.nim(895) semOverloadedCallAnalyseEffects
/home/deech/Nim/compiler/semcall.nim(582) semOverloadedCall
/home/deech/Nim/compiler/semcall.nim(540) semResolvedCall
/home/deech/Nim/compiler/seminst.nim(384) generateInstance
/home/deech/Nim/compiler/seminst.nim(133) instantiateBody
/home/deech/Nim/compiler/semexprs.nim(1847) semProcBody
/home/deech/Nim/compiler/semexprs.nim(3103) semExpr
/home/deech/Nim/compiler/semstmts.nim(2408) semStmtList
/home/deech/Nim/compiler/semexprs.nim(3115) semExpr
/home/deech/Nim/compiler/semexprs.nim(1838) semReturn
/home/deech/Nim/compiler/semexprs.nim(1814) semAsgn
/home/deech/Nim/compiler/sem.nim(106) fitNode
/home/deech/Nim/compiler/types.nim(1648) typeMismatch
/home/deech/Nim/compiler/msgs.nim(571) liMessage
/home/deech/Nim/compiler/msgs.nim(433) handleError
/home/deech/Nim/compiler/msgs.nim(414) quit
FAILURE
```

From the second line of the output it seems the type system thinks the type of `o.j.get` should be `F2[int]` instead of `F3[int]`. If I change `echo typeof(o.j.get)` to `echo typeof(o.j)` it compiles and correctly prints the type as `Option[F3[int]]`. If I remove any of the fields above `j` it works, if I move `j` to an earlier slot in the object it also works.

I tried to diagnose it for a few hours and failed but this is currently blocking my project so any help is appreciated.

### Nim Version

```
Nim Compiler Version 1.7.1 [Linux: amd64]
Compiled at 2022-09-05
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: 8dcf367e5223ae26b57c9bbfaec6e70ac14bb820
active boot switches: -d:release
```

### Current Standard Output Logs

```shell
Included in `What happened?`
```

### Expected Standard Output Logs

```shell
Included in `What happened?`
```

**Comment by beef331** [comment]
This is likely a generic cache bug, given that the error message is literally "A is not the same type as A" a workaround is to use templates instead:
```nim
import std/[options]

template F1*[T]: auto = (proc(self:T,callback:proc(self:T)))
template F2*[T]: auto = (proc(self:T,callback:proc(self:T)))
template F3*[T]: auto = (proc(self:T,callback:proc(self:T)))
  
type
  Functions*[T] = object
    a*: Option[F1[T]]
    b*: Option[F1[T]]
    c*: Option[F1[T]]
    d*: Option[F1[T]]
    e*: Option[F1[T]]
    f*: Option[F1[T]]
    g*: Option[F1[T]]
    h*: Option[F1[T]]
    i*: Option[F2[T]]
    j*: Option[F3[T]]

var o : Functions[int]

static:
  echo typeof(o.j.get)
```

**Comment by deech** [comment]
Interesting, I didn't know you could use templates like that. Unfortunately this isn't a good workaround for me because I do a fair amount of type introspection into the object.

**Comment by deech** [comment]
I tried to debug this for a few more hours and I'm not getting anywhere. Somewhere the `val` of the `Option` of the `j*` field is being instantiated with `F2[int]` instead of `F3[int]` and the instantiation code is a little hard to follow. @Araq this bug seems at least severe.

**Comment by deech** [comment]
I've contained a failing case further:

```nim
type
  AProc = proc | ref | ptr
  MyOption[T] = object
    when T is AProc:
      val: T

proc get[T](o:MyOption[T]):T = o.val

type
  F3*[T] = proc(self:T,callback:proc(self:T))
  F1*[T] = proc(self:T,callback:proc(self:T))
  F2*[T] = proc(self:T,callback:proc(self:T))

type
  Functions*[T] = object
    a*: MyOption[F1[T]]
    b*: MyOption[F1[T]]
    c*: MyOption[F1[T]]
    d*: MyOption[F1[T]]
    e*: MyOption[F1[T]]
    f*: MyOption[F1[T]]
    g*: MyOption[F1[T]]
    h*: MyOption[F1[T]]
    i*: MyOption[F2[T]]
    j*: MyOption[F3[T]]

var o : Functions[int]

static:
  echo typeof(o.j.get)
```

**Comment by beef331** [comment]
Given that it's erroring the same types are not equivalent i'd wager the issue is somewhere inside of https://github.com/nim-lang/Nim/blob/devel/compiler/semtypinst.nim#L327

**Comment by deech** [comment]
I have further reduced it, this prints `F49[int]` when it should print `F50[int]`:

```nim
type
  O[T] = object
    val: T

type
  F0*[T] = string
  F1*[T] = string
  F2*[T] = string
  F3*[T] = string
  F4*[T] = string
  F5*[T] = string
  F6*[T] = string
  F7*[T] = string
  F8*[T] = string
  F9*[T] = string
  F10*[T] = string
  F11*[T] = string
  F12*[T] = string
  F13*[T] = string
  F14*[T] = string
  F15*[T] = string
  F16*[T] = string
  F17*[T] = string
  F18*[T] = string
  F19*[T] = string
  F20*[T] = string
  F21*[T] = string
  F22*[T] = string
  F23*[T] = string
  F24*[T] = string
  F25*[T] = string
  F26*[T] = string
  F27*[T] = string
  F28*[T] = string
  F29*[T] = string
  F30*[T] = string
  F31*[T] = string
  F32*[T] = string
  F33*[T] = string
  F34*[T] = string
  F35*[T] = string
  F36*[T] = string
  F37*[T] = string
  F38*[T] = string
  F39*[T] = string
  F40*[T] = string
  F41*[T] = string
  F42*[T] = string
  F43*[T] = string
  F44*[T] = string
  F45*[T] = string
  F46*[T] = string
  F47*[T] = string
  F48*[T] = string
  F49*[T] = string
  F50*[T] = string

type
  Fields*[T] = object
    f0*: O[F0[T]]
    f1*: O[F1[T]]
    f2*: O[F2[T]]
    f3*: O[F3[T]]
    f4*: O[F4[T]]
    f5*: O[F5[T]]
    f6*: O[F6[T]]
    f7*: O[F7[T]]
    f8*: O[F8[T]]
    f9*: O[F9[T]]
    f10*: O[F10[T]]
    f11*: O[F11[T]]
    f12*: O[F12[T]]
    f13*: O[F13[T]]
    f14*: O[F14[T]]
    f15*: O[F15[T]]
    f16*: O[F16[T]]
    f17*: O[F17[T]]
    f18*: O[F18[T]]
    f19*: O[F19[T]]
    f20*: O[F20[T]]
    f21*: O[F21[T]]
    f22*: O[F22[T]]
    f23*: O[F23[T]]
    f24*: O[F24[T]]
    f25*: O[F25[T]]
    f26*: O[F26[T]]
    f27*: O[F27[T]]
    f28*: O[F28[T]]
    f29*: O[F29[T]]
    f30*: O[F30[T]]
    f31*: O[F31[T]]
    f32*: O[F32[T]]
    f33*: O[F33[T]]
    f34*: O[F34[T]]
    f35*: O[F35[T]]
    f36*: O[F36[T]]
    f37*: O[F37[T]]
    f38*: O[F38[T]]
    f39*: O[F39[T]]
    f40*: O[F40[T]]
    f41*: O[F41[T]]
    f42*: O[F42[T]]
    f43*: O[F43[T]]
    f44*: O[F44[T]]
    f45*: O[F45[T]]
    f46*: O[F46[T]]
    f47*: O[F47[T]]
    f48*: O[F48[T]]
    f49*: O[F49[T]]
    f50*: O[F50[T]]

var fs : Fields[int]

static:
  echo typeof(fs.f50.val)
```
